docs/proposals/security-context-constraints.md

```diff
@@ -0,0 +1,292 @@
+1.  Provide the ability for users and infrastructure components to run pods with elevated privileges
+on behalf of another user or within a namespace where privileges are more restrictive.
+1.  Secure the ability to reference elevated permissions or to change the constraints under which
+a user runs.
+
+### Use Cases
+
+Use cases 1:
+As an administrator, I can create a namespace for a person that can't create privileged containers
+AND enforces that the UID of the containers is set to a certain value
+
+Use cases 2:
+A build controller should be able to create a privileged pod in a namespace, but regular users
+can't create privileged pods AND regular users can't exec into that pod.
+
+Use cases 3:
+As a cluster administrator, I can allow a given namespace (or service account) to create privileged
+pods or to run root pods
+
+Use case 4:
+As a cluster administrator, I can allow a project administrator to control the security contexts of
+pods and service accounts within a project
+
+
+### Requirements
+
+1.  Provide a set of restrictions that a security context operates under a new object will be
+introduced called SecurityContextConstraints.
+1.  SecurityContextConstraints may also exist outside of a namespace.
+1.  A cluster default service account will be created for the default namespace and be available to
+admission controllers via configuration.
+1.  User information must be available to admission controllers.
+1.  Some authorizers may restrict a user’s ability to reference a service account.  Systems requiring
+the ability to secure service accounts on a user level will be able to add a policy that enables
+referencing specific service accounts themselves.
+1.  Admission control must validate the creation of Pods, PodTemplates, and Replication controllers
+against the allowed set of constraints.
+
+### Design
+
+#### Model
+
+The model for a security context constraint is a combination of the security context constraint
+itself and cluster wide strategies enforced by an allocator.  The separation of concerns falls into
+two distinct categories: permissions that the pod can request and runtime parameters the cluster
+admin controls.
+
+An example of a permission that can be requested is to run a pod as privileged.  This is set on a
+pod by pod basis.  A cluster wide setting may be the allocation of user ids that pods must run under.
+The cluster administrator may decide that no pod can run as the root user in their cluster and
+provide a block of UIDs that the allocator can use to distribute amongst namespaces.
+
+A security context constraint object exists in the root scope, outside of a namespace.  The
+security context constraints can reference users, groups, or a service account that is allowed
+to operate under the constraints.
+
+A one to one relationship will be maintained for a security context constraint and a service account.
```

Comment:
I'm returning to the idea that unioning is not a good idea as noted above. Since the strategies really belong to the SCC and not cluster wide it seems to make more sense to have a direct match or no match on an SCC rather than a union so we know exactly which strategy to validate against.